runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,346 @@
+              end
+              append(vim.tbl_flatten(parsed_choices))
+              in_braces = false
+              brace_val = ""
+            elseif char == "]" then
+              append({"["..bracket_val.."]"})
+              in_brackets = false
+              bracket_val = ""
+            elseif char == "?" then
+              append({non_path_sep})
+            elseif char == "*" then
+              append({non_path_sep.."-"})
+            else
+              append({escape(char)})
+            end
+          end
+        end
+      end
+
+      if not last_seg and (segments[i+1] ~= "**" or i+1 < #segments) then
+        append({path_sep})
+      end
+    end
+  end
+
+  return patterns
+end
+
+---@private
+--- Implementation of LSP 3.17.0's pattern matching: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#pattern
+--- Modeled after VSCode's implementation: https://github.com/microsoft/vscode/blob/0319eed971719ad48e9093daba9d65a5013ec5ab/src/vs/base/common/glob.ts#L509
+---
+---@param pattern string|table The glob pattern (raw or parsed) to match.
+---@param s string The string to match against pattern.
+---@return bool Whether or not pattern matches s.
+local function match(pattern, s)
+  if type(pattern) == "string" then
+    pattern = parse(pattern)
+  end
+  -- Since Lua's built-in string pattern matching does not have an alternate
+  -- operator like '|', `parse` will construct one pattern for each possible
+  -- alternative. Any pattern that matches thus matches the glob.
+  for _, p in ipairs(pattern) do
+    if s:match("^"..p.."$") then return true end
+  end
+  return false
+end
+
+---@private
+--- Joins filepath elements by platform-specific separator.
+---
+---@param args ... The path elements. The first element must be absolute.
+---@return string The joined path.
+local function filepath_join(...)
+  local dir = ...
+  return table.concat({...}, dir:match('^([a-zA-Z]:)(.*)') and "\\" or "/")
+end
+
+-- Cache of libuv handles per directory, per LSP client, per registration ID.
+local watched_dirs = {}
+
+local change_queue = {}
+local change_cache = {}
+local queue_timers = {}
+
+---@private
+--- Creates callbacks invoked on watched file events.
+---
+---@param dir string Absolute path to the watched directory.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+---@return function The callback invoked on watched file events.
+local function get_callback(dir, client_id, reg_id)
+  return function(filename, events)
+    local path = filepath_join(dir, filename)
+
+    local type = events.change and protocol.FileChangeType.Changed or 0
+    if events.rename then
+      local _, err, errname = uv.fs_stat(path)
+      if errname == "ENOENT" then
+        type = protocol.FileChangeType.Deleted
+      else
+        assert(not err, err)
+        type = protocol.FileChangeType.Created
+      end
+    end
```

Comment:
This translates the libuv event into an LSP change type. There is a race here where if a file is created and deleted in quick succession, the sequence of changes might contain two "deleted" events if the `stat` call for the actual "created" event happens after the file is already gone.

I wonder how big of an issue this is depending on how robust language servers are to these kinds of inconsistencies.

Reply:
Some kind of debouncing to avoid that could probably help, but that can be done as a follow up too.

Reply:
The only debouncing here so far is skipping consecutive changes of the same type for the same file, but I agree there are probably opportunities to be smarter about pruning events.